doc/code-walkthrough.md

```diff
@@ -0,0 +1,169 @@
+# High-level code walkthrough
+
+## Code inventory and testing strategy
+
+The Shields codebase is divided into several parts:
+
+1.  The frontend (about 7% of the code)
+    1. [`frontend`][frontend]
+2.  The badge renderer (which is available as an npm package)
+    1.  [`gh-badges`][gh-badges]
+3.  The base service classes (about 8% of the code, and probably the most important
+    code in the codebase)
+    1.  [`core/base-service`][base-service]
+4.  The server code and a few related odds and ends
+    1.  [`core/server`][server]
+5.  Helper code for token pooling and persistence (used to avoid GitHub rate limiting)
+    1.  [`core/token-pooling`][token-pooling]
+6.  Service common helper functions (about 7% of the code, and fairly important
+    since it’s shared across much of the service code)
+    1.  `*.js` in the root of [`services`][services]
+7.  The services themselves (about 80% of the code)
+    1.  `*.js` in the folders of [`services`][services]
+8.  The badge suggestion endpoint (Note: it's tested as if it’s a service.)
+    1.  [`services/suggest.js`][suggest]
+
+[frontend]: https://github.com/badges/shields/tree/master/frontend
+[gh-badges]: https://github.com/badges/shields/tree/master/gh-badges
+[base-service]: https://github.com/badges/shields/tree/master/core/base-service
+[server]: https://github.com/badges/shields/tree/master/core/server
+[token-pooling]: https://github.com/badges/shields/tree/master/core/token-pooling
+[services]: https://github.com/badges/shields/tree/master/services
+[suggest]: https://github.com/badges/shields/tree/master/services/suggest.js
```

Comment:
Should this be the `suggest` directory? https://github.com/badges/shields/tree/master/services/suggest

Reply:
Oops, it's just not moved there yet.

Reply:
I opened #2873. Will update this to point to the pre-move file path in the meantime.